keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+    - [<code>--node-ip</code> and Address Sorting](#-and-address-sorting)
+    - [Updating Current Node.Status.Addresses Users](#updating-current-nodestatusaddresses-users)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+There are problems with the handling of `Node.Status.Addresses` that
+currently break or complicate some IPv6 and dual-stack scenarios.
+
+Additionally, `Node.Status.Addresses` behaves inconsistently between
+bare metal, legacy cloud providers, and external cloud providers and
+generally has unclear and inconsistent semantics. It also has a broken
+definition, resulting in it needing to have a note attached to it in
+the documentation warning users to not use it with strategic merge
+patch.
+
+To resolve these problems, this KEP proposes deprecating the existing
+`Node.Status.Addresses` field and replacing it with two new fields,
+`Node.Status.Hostname` and `Node.Status.IPs`, with clearer semantics
+and more consistent and useful behavior.
```

Comment:
I tend to agree with the direction of https://github.com/kubernetes/enhancements/pull/1665/files#r426330540 and https://github.com/kubernetes/enhancements/pull/1665/files#r426153433 ... 

Since we have to continue supporting `.status.addresses` indefinitely (until a hypothetical unplanned v2 node) by populating it (to avoid breaking current readers) and reading it (to avoid breaking current external cloud providers), I have a hard time seeing how adding a second field containing the same data makes the situation better. If anything, I anticipate it making the situation more complicated, since interactions between the two fields would then need to be considered.

Even if there were a new field, strategic merge patch still should not be used to update both the old and new field at the same time. This is due both to the broken strategic merge patch definition on the current field, and because strategic merge patch is (ironically) not the strategic direction apimachinery is headed for patching (server-side-apply patches are).

If there are clients that set the existing field in problematic ways, would it be simpler to adjust them to be better behaved with the existing field than to require them to also set a new field? If there are poorly behaved clients it is impossible to change, won't it also be impossible to change them to use the new field proposed here?

What would it look like to improve .status.addresses?
* Describe how to replace it (merge patch, jsonpatch, apply patch)
* Document how kube-apiserver and kube-proxy determine an address (type, order, family, etc)
* Defining specific IPv4/IPv6 address types
* Documenting the in-tree uses for address info, along with the lookup algorithm used to determine an address for each purpose
* Creating type- and family-aware helpers to set and retrieve data from the list for family-specific purposes
* ... something else?